### DISCUSSION TITLE
Support for Debugging Neko Simulations in VS Code

### DISCUSSION THREAD

**Discussion by SachinBM-CE** [ExtremeFLOW/neko#1638] [Q&A] [answered]
I want to use Visual Studio Code (VS Code) as my IDE for working with Neko, particularly to leverage its debugging capabilities. However, setting this up does not seem straightforward to me. When attempting to debug, VS Code prompts me to select a debugger and create `launch.json` configuration files.

My environment includes Windows Subsystem for Linux (WSL) and Open MPI. I aim to debug code using MPI and CUDA.

I request your support on:
- Recommendations for a compatible debugger for Neko / Modern-Fortran development
- Example configurations for `launch.json` to handle common Fortran debugging scenarios in Neko

If VS Code is not ideal for this application, I’d appreciate suggestions for alternative IDEs or tools better suited for Fortran, MPI, and CUDA debugging workflows.

**Comment by timofeymukha** [comment]
@timfelle might be able to help. I believe I managed to do this on one machine for the CPU, and incidentally I will have access to it next week, so I can share the setup.

**Comment by timfelle** [comment, marked as answer]
Hi, Yes I have indeed dabbled a bit in this.

Extensions (I think these are enough, its been a while):
- fortran-lang.linter-gfortran
- ms-vscode.cpptools

An example of a `launch.json` is given below.
I would suggest you look into the "preLaunchTask" which would go into a `tasks.json` if you want to compile the code before the debug is launched.

I have done so, however, I am not working directly on neko but through an extension library for Topology Optimization, hence the build system is quite different from Neko itself.

But the gist of the preLaunchTask is to define something like the second dropdown, where you do your general setup of Neko. And then also run the "makeneko" script for the specific example you want to use for debugging if you have user defined code in there.

I have a few notes in the Neko-TOP repository on Cuda and WSL if that is useful to you:
https://extremeflow.github.io/neko-top/dc/dec/compilation.html#autotoc_md15

Another note:
I have not been able to get debugging features of the cuda kernels themselves, since they are launched as external code through an interface in Fortran. But such is life with these mixed language systems. If you find a solution to it that would be great.

<details>
<summary>My `launch.json`</summary>

```.json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(gdb) Fortran Neko-Top Debugging",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/bin/neko",
            "args": [
                "cylinder.case"
            ],
            "stopAtEntry": false,
            "cwd": "${workspaceFolder}/examples/cylinder",
            "environment": [
                {
                    "name": "NEKO_LOG_LEVEL",
                    "value": "10"
                },
                {
                    "name": "NEKO_LOG_FILE",
                    "value": "neko.log"
                },
                {
                    "name": "LD_LIBRARY_PATH",
                    "value": "/path/to/json-fortran/lib"
                }
            ],
            "externalConsole": false,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            // "preLaunchTask": "Initialize Debug",
        }
    ]
}
```
</details>

<details>
<summary>tasks.json</summary>

```.json
{
    "version": "2.0.0",
    "tasks": [
        {
            "detail": "Prepare the Neko-TOP debug environment",
            "type": "shell",
            "label": "Initialize Debug",
            "command": "${workspaceFolder}/configure",
            "args": [
                "--prefix=${workspaceFolder",
                "FC=/some/compiler"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "problemMatcher": [ ],
            "dependsOn": [
                "Prepare environment",
            ],
            "presentation": {
                "echo": true,
                "reveal": "never",
                "revealProblems": "onProblem",
                "focus": false,
                "panel": "shared",
                "showReuseMessage": true,
                "clear": false,
                "close": true
            }
        }
    ]
}
```
</details>

**Comment by SachinBM-CE** [comment]
Hi @timofeymukha & @timfelle,

Thank you so much for the detailed responses and the sample configurations! Your guidance was invaluable in setting it up.

Following are the same files for the case I'm running & as per my system settings:
[launch.json](https://github.com/user-attachments/files/18100495/launch.json)
[tasks.json](https://github.com/user-attachments/files/18100497/tasks.json)

I assume that the above does the following:
```
makeneko turb_channel.f90
./neko turb_channel.case
```

The above is working perfectly fine. It is opening the debugger & stops at breakpoints.

However, I’d like to include mpirun in the workflow to run the case with MPI, as follows:
`mpirun -np 4 ./neko turb_channel.case`

For this I tried to modify launch.json, as follows:
```
"args": [
                "-np", "4", "turb_channel.case"
            ]
```

Please let me know how do I specify `mpirun` in `launch.json`?

**Comment by timfelle** [comment]
So in general i think VSCode will fall short if you wish to do MPI based debugging.

VSCode calls gdb internally and judging from: https://stackoverflow.com/questions/329259/how-do-i-debug-an-mpi-program I think you would need to launch a gdb session for each of the mpi ranks, which I don't see how you would do in the vscode system.

**Comment by njansson** [comment]
Let's convert this into a discussion, feel free to continue the discussion in the converted thread under Q&A
